cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+// or a template. To scale down the cluster, delete specific instances of
+// Machines and the underlying Nodes will be unregistered/deprovisioned.
+// Separate provider-specific controllers will watch Machine objects they can
+// act on (like a GCE cloud controller watching for only Machines destined for
+// GCE) and take the appropriate actions.
+//
+// Any updates to the MachineSpec will be actuated to change the Node in
+// place or replace the Node with one conforming to the spec. In this model,
+// the fact that a controller is able to upgrade a node via in-place upgrades
+// or via a cloud replacement is an implementation detail without API controls.
+//
+// It is recommended, but not required, that provider-specific controllers add
+// finalizers to Machine objects so that they can be triggered on deletion to
+// release the necessary external resources, reporting any errors encountered.
+type Machine struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   MachineSpec
+	Status MachineStatus
+}
+
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
+	// when creating the Node.
+	// +optional
+	metav1.ObjectMeta
+
+	// Provider-specific serialized configuration to use during node
+	// creation. It is recommended that providers maintain their own
+	// versioned API types that should be serialized/deserialized from this
+	// field, akin to component config.
+	// +optional
+	ProviderConfig string
+
+	// A list consisting of "Master" and/or "Node".
+	//
+	//                 +-----------------------+------------------------+
+	//                 | Master present        | Master absent          |
+	// +---------------+-----------------------+------------------------|
+	// | Node present: | Install control plane | Join the cluster as    |
+	// |               | and be schedulable    | just a node            |
+	// |---------------+-----------------------+------------------------|
+	// | Node absent:  | Install control plane | Invalid configuration  |
+	// |               | and be unscheduleable |                        |
+	// +---------------+-----------------------+------------------------+
+	Roles []string
+
+	// Versions of key software to use.
+	// +optional
+	Versions MachineVersionInfo
+
+	// To populate in the associated Node for dynamic kubelet config. This
+	// field already exists in Node, so any updates to it in the Machine
+	// spec will be automatially copied to the linked NodeRef from the
+	// status. The rest of dynamic kubelet config support should then work
+	// as-is.
+	// +optional
+	ConfigSource *corev1.NodeConfigSource
+}
+
+type MachineStatus struct {
+	// If the corresponding Node exists, this will point to its object.
+	// +optional
+	NodeRef *api.ObjectReference
+
+	// When was this status last observed
+	// +optional
+	LastUpdated metav1.Time
```

Comment:
Let's make clear that this is last-transition, rather than a heartbeat.   (If we hadn't done the node heartbeat, I'd wager we'd be at 10k nodes by now...)

Reply:
Will do.